codalab/worker_manager/slurm_batch_worker_manager.py

```diff
@@ -0,0 +1,241 @@
+            type=str,
+            help='Path to the file containing the username and '
+            'password for logging into the CodaLab worker '
+            'each on a separate line. If not specified, the '
+            'the worker will fail to start.',
+        )
+
+    def __init__(self, args):
+        super().__init__(args)
+        self.username = self.args.user if self.args.user else getpass.getuser()
+        # A set of newly submitted job id to keep tracking worker status, as worker might not be created right away.
+        self.submitted_jobs = set()
+
+    def get_worker_jobs(self):
+        """
+        Return a list of workers in RUNNING and PENDING state.
+        The current Slurm version on NLP cluster is 17.11.13-2, which doesn't have rest api provided.
+        """
+        # Documentation can be found at https://slurm.schedmd.com
+        # Since allocating resource for a worker may take a while, we periodically check
+        # for worker status and remove those workers that failed at starting phase.
+        failed_jobs = set()
+        for job_id in self.submitted_jobs:
+            job_acct = self.run_command(['sacct', '-j', job_id, '--format', 'state'])
+            if 'FAILED' in job_acct:
+                failed_jobs.add(job_id)
+        self.submitted_jobs = self.submitted_jobs - failed_jobs
+
+        # Get all the Slurm workers that are submitted by SlurmWorkerManager and owned by the current user.
+        # Returning result will be in the following format:
+        # JOBID:STATE (header won't be included with "--noheader" option)
+        # 1478828:PENDING
+        # 1478830:PENDING
+        jobs = self.run_command(['squeue', '-u', self.username, '--format', '%i:%T', '--noheader'])
+        jobs = jobs.strip().split()
+        logger.info(
+            'Workers: {}'.format(
+                ' '.join(job for job in jobs if job in self.submitted_jobs) or '(none)'
+            )
+        )
+
+        # Get all the PENDING and RUNNING jobs that are owned by the current user.
+        # Returning result will be in the following format:
+        # JOBID (header won't be included with "--noheader" option)
+        # 1478828
+        # 1478830
+        running_jobs = self.run_command(
+            ['squeue', '-u', self.username, '-t', 'PENDING,RUNNING', '--format', '%i', '--noheader']
+        )
+        running_jobs = running_jobs.strip().split()
+
+        # Return jobs in PENDING (waiting for resources) and RUNNING (actively running)
+        # states here as we don't want to create more workers that we actually need.
+        return [WorkerJob(job) for job in running_jobs if job in self.submitted_jobs]
+
+    def start_worker_job(self):
+        """
+        Start a CodaLab Slurm worker that submits batch job to Slurm
+        """
+        worker_id = uuid.uuid4().hex
+        # Set user's local home directory as worker directory
+        work_dir = Path(str(Path.home()), "slurm-batch-scratch/{}".format(worker_id))
```

Comment:
I think this should be configurable somehow—if we leave this as-is for the Stanford NLP cluster, we'd be writing to NFS instead of the worker's local disk, which would be pretty slow (and possibly impact other users).

Something to keep in mind: if we all put them into a fixed path (say, `hostname/scr1`, like the default `nlp-cl-worker` script does, it's possible that that folder could become full and a bunch of jobs would fail. Personally, I use a script `get-scr-with-most-space` that returns the `hostname/scr*` with the most space, and set work-dir to `$(get-scr-with-most-space)/$USER-cl-worker`...not sure how that'd fit in here, but it'd be nice to spread around the usage somehow.

Reply:
hmm, disk usage could be problematic here. I think spreading around disk usage will slow down the issue, but it will happen eventually at some point. I guess my question is if `homename/scr*` a shared disk space that can be used for all users when they are running the Slurm worker manager? Not sure if we can allocate some space to save the worker data.

Reply:
Maybe we should allow workers to delete their `work-dir` upon exit?

Reply:
Yeah, that sounds like a good idea. In addition to this, I am also wondering maybe we can pass the `work-dir` an option to let users choose where they would want to put their data instead of hardcoded there?

Reply:
yup, that sounds great! Except it'd be a `work-dir-prefix` in this case, right?